### PR TITLE
Only run providers that are referenced in the policy

### LINKED ISSUE #3609: Run only providers that have references in the policy
**Describe the enhancement:**

Currently the Elastic Agent runs all of its providers all the time. Context providers are rather low overhead, but dynamic providers are not and when they are not even being used its adding load for no benefit to the running policy.

We should adjust the Elastic Agent to only run providers that are referenced in the policy. Variable referencing is always prefixed with a provider name aka `${kubernetes.*}` that means its rather easy to identify which providers should be enabled based on what is in the policy.

This also removes the need to add any configuration to Fleet to disable or enable a provide for the Elastic Agent. If the policy from Fleet doesn't use a provider then the Elastic Agent will not run it.

**Describe a specific use case for the enhancement or feature:**

Reduce load of an Elastic Agent, prevents it from trying to communicate with Docker or Kubernetes in the case that its not even running a Docker or Kubernetes inputs in the policy.

**What is the definition of done?**

I propose the following change:

1. Parse the policy into the AST.
2. Read the AST to determine the providers that are referenced.
3. Start/stop the providers that are referenced or no longer referenced.

### LINKED ISSUE #4648: Elastic Agent is waiting for initial configuration forever when all providers are disabled
In case of no enabled providers, the elastic agent stalls forever.
It seems to be a bug here:
https://github.com/elastic/elastic-agent/blob/0c7212f2d92021a9e008de4abe362d0c77f78638/internal/pkg/composable/controller.go#L188-L203
where there is no way to break to `DEBOUNCE` if no provider is updating.

A possible solution could be based on the providers config length or the waiting group which should return immediately in this case since if there are no providers it is equal to 0.
https://github.com/elastic/elastic-agent/blob/0c7212f2d92021a9e008de4abe362d0c77f78638/internal/pkg/composable/controller.go#L128

The bug was discovered as part of the work on agentless controller. Currently we use a [workaround](https://github.com/elastic/agentless-controller/pull/343) to solve this issue.

Bug details:
- Version: 8.13.2
- Operating System: all
- Steps to Reproduce:

Setup (`agent-bug` is the directory name which shows up before every command)
```
➜  test cd agent-bug
➜  agent-bug docker pull docker.elastic.co/beats/elastic-agent:8.13.2
8.13.2: Pulling from beats/elastic-agent
c93e5d1261d3: Pull complete
9204e5b4f4d9: Pull complete
77db57972e9d: Pull complete
5a09faecb150: Pull complete
ef1d475c705b: Pull complete
9be0bc4d4489: Pull complete
97bac83776bc: Pull complete
594c586edc1b: Pull complete
b45e1922fc73: Pull complete
a9c1a4bc09dd: Pull complete
83a50bca82ec: Pull complete
4ca545ee6d5d: Pull complete
Digest: sha256:1b1346f6228c4cfcc8bd6b05e0eb24f15bcfd616935d0f2fffe0754d7d3fe31b
Status: Downloaded newer image for docker.elastic.co/beats/elastic-agent:8.13.2
docker.elastic.co/beats/elastic-agent:8.13.2
```
Get the original config file
```
➜  agent-bug docker run --rm -d --name elastic-agent docker.elastic.co/beats/elastic-agent:8.13.2 container
76af237adeb219eb7591c6b7647c3bfe523e4d73121bf1677388720b79e29d85
➜  agent-bug docker cp elastic-agent:/usr/share/elastic-agent/elastic-agent.yml ./
➜  agent-bug docker stop elastic-agent
```
Modify it to disable all providers
```
➜  agent-bug cat <<EOF >> elastic-agent.yml
providers:
  agent:
    enabled: false
  docker:
    enabled: false
  env:
    enabled: false
  host:
    enabled: false
  kubernetes:
    enabled: false
  kubernetes_leaderelection:
    enabled: false
  kubernetes_secrets:
    enabled: false
  local:
    enabled: false
  local_dynamic:
    enabled: false
  path:
    enabled: false
EOF
```
Start the agent with the modified config
```
➜  agent-bug docker run --rm -d --name elastic-agent -v ./elastic-agent.yml:/usr/share/elastic-agent/elastic-agent.yml docker.elastic.co/beats/elastic-agent:8.13.2 container
decb5e6e6d443e11d228a9ff32e8a9ad3ed78f466b7c830e85aec0a3818b9aa4
```
Enter the container
```
➜  agent-bug docker exec -it elastic-agent /bin/bash
```
Agent stuck on waiting for initial configuration
```
elastic-agent@decb5e6e6d44:~$ elastic-agent status
┌─ fleet
│  └─ status: (STOPPED) Not enrolled into Fleet
└─ elastic-agent
   └─ status: (STARTING) Waiting for initial configuration and composable variables
```
... waiting ...
```
elastic-agent@decb5e6e6d44:~$ elastic-agent status
┌─ fleet
│  └─ status: (STOPPED) Not enrolled into Fleet
└─ elastic-agent
   └─ status: (STARTING) Waiting for initial configuration and composable variables
```
Agent configuration (for some reason inspect stalls so I have to kill it)
```
elastic-agent@decb5e6e6d44:~$ elastic-agent inspect
agent:
  logging:
    to_stderr: true
inputs:
- data_stream.namespace: default
  id: unique-system-metrics-input
  streams:
  - data_stream.dataset: system.cpu
    metricsets:
    - cpu
  - data_stream.dataset: system.memory
    metricsets:
    - memory
  - data_stream.dataset: system.network
    metricsets:
    - network
  - data_stream.dataset: system.filesystem
    metricsets:
    - filesystem
  type: system/metrics
  use_output: default
outputs:
  default:
    hosts: http://elasticsearch:9200
    password: changeme
    preset: balanced
    type: elasticsearch
    username: elastic
providers:
  agent:
    enabled: false
  docker:
    enabled: false
  env:
    enabled: false
  host:
    enabled: false
  kubernetes:
    enabled: false
  kubernetes_leaderelection:
    enabled: false
  kubernetes_secrets:
    enabled: false
  local:
    enabled: false
  local_dynamic:
    enabled: false
  path:
    enabled: false

^CError: could not load agent info: could not get agent info from store: failed to load from ioStore: failed to ensure key during encrypted disk store Load: could not get agent key: failed to acquire exclusive lock: /usr/share/elastic-agent/state/vault/.lock, err: context canceled
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.13/fleet-troubleshooting.html
elastic-agent@decb5e6e6d44:~$ exit
```
Cleanup
```
➜  agent-bug docker stop elastic-agent
elastic-agent
```

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

This updates the Elastic Agent to only run composable providers if they are referenced in the policy.

## Why is it important?

This change provides the following benefits:

- Reduces the memory and load of the Elastic Agent, by only running the providers it references.
- No longer requires access to the kubernetes API or resources just to run an Elastic Agent (as long as its not using any ${kubernetes*` references in the policy).
- Fixes an issue where if no providers are running that a set of empty variables is passed to the coordinator.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~ (covered by unit tests)

## Disruptive User Impact

None. As it still runs all those that are referenced from the user.

## How to test this PR locally

Run the elastic-agent container with just the system integration and notice that it doesn't require access to the Kubernetes API any more.

## Related issues

- Closes #3609 
- Closes #4648

### CODE REVIEW

File: internal/pkg/agent/transpiler/ast.go [hunk 1]
```diff
@@ -982,6 +1064,11 @@ func attachProcessors(node Node, processors Processors) Node {
 
 // Lookup accept an AST and a selector and return the matching Node at that position.
 func Lookup(a *AST, selector Selector) (Node, bool) {
+	// Be defensive and ensure that the ast is usable.
+	if a == nil || a.root == nil {
+		return nil, false
+	}
```
Comment (swiatekm) [inline internal/pkg/agent/transpiler/ast.go:1070]: We should probably write some tests for these kinds of cases. Not necessarily in this PR, but I recall recently adding similar conditions.

File: internal/pkg/composable/controller.go [hunk 2]
```diff
@@ -192,8 +171,13 @@ func (c *controller) Run(ctx context.Context) error {
...
 			select {
 			case <-ctx.Done():
-				cleanupFn()
 				return ctx.Err()
+			case observed := <-c.observedCh:
+				c.handleObserved(localCtx, &wg, stateChangedChan, fetchContextProviders, observed)
+				t.Reset(100 * time.Millisecond)
+				c.logger.Debugf("Observed state changed for composable inputs; debounce started")
+				drainChan(stateChangedChan)
+				break DEBOUNCE
```
Comment (swiatekm) [inline internal/pkg/composable/controller.go:180]: This is so we always update vars after `Observe` is called? If so, we should make sure we only emit if anything changed.

File: internal/pkg/composable/controller.go [hunk 3]
```diff
@@ -35,18 +35,24 @@ type Controller interface {
...
-	dynamicProviders map[string]*dynamicProviderState
+	logger          *logger.Logger
+	ch              chan []*transpiler.Vars
+	observedCh      chan map[string]bool
+	errCh           chan error
+	restartInterval time.Duration
+
+	managed          bool
+	contextProviders map[string]contextProvider
+	dynamicProviders map[string]dynamicProvider
```
Comment (swiatekm) [inline internal/pkg/composable/controller.go:52]: These are more like provider *factories* rather than providers themselves, right?

**Review by swiatekm — COMMENTED** [review]
Looks good overall, had some questions about the design and intended control flow.

File: internal/pkg/agent/transpiler/ast.go [hunk 4]
```diff
@@ -58,6 +58,9 @@ type Node interface {
 	// Hash compute a sha256 hash of the current node and recursively call any children.
 	Hash() []byte
 
+	// Vars return a list of referenced vars.
+	Vars([]string) []string
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/transpiler/ast.go:62]: Correct.

File: internal/pkg/composable/controller.go (hunk 3 above)
Comment (blakerouse) [reply internal/pkg/composable/controller.go:52]: Correct.

File: internal/pkg/agent/transpiler/ast.go (hunk 1 above)
Comment (blakerouse) [reply internal/pkg/agent/transpiler/ast.go:1070]: I can add a test.

File: internal/pkg/composable/controller.go (hunk 2 above)
Comment (blakerouse) [reply internal/pkg/composable/controller.go:180]: This should only happen when the policy changes which should not be often, at least not as often compared to a variable changing in a dynamic provider like kubernetes. I will update it to only emit if something actually changed.

File: internal/pkg/agent/transpiler/ast.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (swiatekm) [reply internal/pkg/agent/transpiler/ast.go:62]: We should add a comment explaining that, imo. It wasn't obvious to me, and I was a bit confused looking at the implementations. Maybe change the name to `VarsInto`?

[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/transpiler/ast.go:63]: I added a comment in latest commit.

File: internal/pkg/agent/transpiler/ast.go [hunk 5]
```diff
@@ -472,6 +500,40 @@ func (s *StrVal) Hash() []byte {
...
+				case *constString:
+					// not a variable
+				case *varString:
+					// found variable add it to the array
+					vars = append(vars, val.Value())
+				}
+			}
+		}
+	}
+	return vars
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/transpiler/ast.go]: I updated this to share the code.

File: internal/pkg/composable/controller.go (hunk 2 above)
Comment (blakerouse) [reply internal/pkg/composable/controller.go]: Updated in latest commit.

File: internal/pkg/composable/controller.go (hunk 3 above)
Comment (blakerouse) [reply internal/pkg/composable/controller.go]: Updated variable naming to make it easier to read.

File: internal/pkg/agent/transpiler/ast.go (hunk 1 above)
Comment (blakerouse) [reply internal/pkg/agent/transpiler/ast.go:1047]: Test added.

**Review by swiatekm — APPROVED** [review]
:+1:

**Comment by cmacknz** [comment]
Nice! We need to update the docs to indicate that providers are no longer enabled by default for the first release that will have this change (8.18 I believe): https://www.elastic.co/guide/en/fleet/current/providers.html#disable-providers-by-default

Some of the individual providers also indicate they are enabled by default like https://www.elastic.co/guide/en/fleet/current/kubernetes_leaderelection-provider.html

CC @kilfoyle

**Pull request merged**